[wasm][build] Build stdlib and test IRGen for WebAssembly

kateinoigakukun opened the pull request:
This patch adds new build products to build and test the Swift stdlib for WebAssembly.
This adds WebAssembly specific stdlib product instead of adding the new target in `stdlib-deployment-targets` because unlike darwin platforms Wasm target has quite different stdlib configuration from host target and there is no way to specify different stdlib configuration for each target in `stdlib-deployment-targets` for now.

This patch adds 1 new dependency `wasi-libc` and 3 new build products:
1. `WASILibc` - wasi-libc is a libc implementation for WebAssembly System Interface (WASI). This product is required to build the compiler runtimes and Swift stdlib.
2. `WasmLLVMRuntimeLibs` - This product builds `./runtimes` directory in llvm-project, which contains compiler-rt, libcxx and libcxxabi. This product is required to build Swift stdlib.
3. `WasmStdlib` - This product builds standalone Swift stdlib for WebAssembly. (Similar to `MinimalStdlib` product, but configured for WebAssembly target)


Currently `WasmStdlib` runs test suite under `test/IRGen` to verify the semi-target specific code in IRGen works and the generated IR can be processed by LLVM WebAssembly backend. Rest of test suites including executable tests are now skipped, but will be enabled after we set up a Wasm runtime in the CI infrastructure and calling convention mismatch issues will be resolved. 

Resolves rdar://116007434 and rdar://116523979.

kateinoigakukun committed:
[wasm][build] Build and test stdlib for WebAssembly

This patch adds a new build products to build and test the Swift stdlib
for WebAssembly.
This adds WebAssembly specific stdlib product instead of adding the new
target in `stdlib-deployment-targets` because unlike darwin platforms
Wasm target has quite different stdlib configuration from host target
and there is no way to specify different stdlib configuration for each
target in `stdlib-deployment-targets` for now.

This patch adds 1 new dependency `wasi-libc` and 3 new build products:
1. `WASILibc` - wasi-libc is a libc implementation for WebAssembly
   System Interface (WASI). This product is required to build the
   compiler runtimes and Swift stdlib.
2. `WasmLLVMRuntimeLibs` - This product builds `./runtimes` directory
   in llvm-project, which contains compiler-rt, libcxx and libcxxabi.
   This product is required to build Swift stdlib.
3. `WasmStdlib` - This product builds standalone Swift stdlib for
   WebAssembly. (Similar to `MinimalStdlib` product, but configured for
   WebAssembly target)

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun commented:
preset= buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun commented:
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
The stdlib compilation crash will be resolved after rebranch will be merged

kateinoigakukun committed:
[wasm][test] Disable a test that depends on swifttailcc

kateinoigakukun commented:
I confirmed it passes with the llvm fix included in the rebranch https://github.com/apple/swift/pull/69239

shahmishal commented on utils/update_checkout/update-checkout-config.json:
```diff
@@ -131,7 +133,8 @@
                 "swift-nio-ssl": "2.15.0",
-                "swift-experimental-string-processing": "swift/main"
+                "swift-experimental-string-processing": "swift/main",
+                "wasi-libc": "wasi-sdk-20"
```
Also add this to `rebranch` and `next` schemes.

shahmishal commented on test/IRGen/async_dynamic_replacement.swift:
```diff
@@ -1,6 +1,8 @@
 // RUN: %target-swift-frontend %s -emit-ir -disable-availability-checking -disable-objc-interop | %FileCheck %s
 
 // REQUIRES: concurrency
+// LLVM does not support swifttailcc for WebAssembly target for now (rdar://116753879)
```
Can we create an issue on GitHub too?

kateinoigakukun commented:
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun committed:
[wasm][build] Skip wasi-libc symbol check step to support LLVM 17

kateinoigakukun commented:
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun committed:
[wasm] Exclude builtin headers from WASILibc system module

stable/20230725 now rejects including builtin headers in modulemap by
default to avoid module cycles on non-Apple platforms. Builtin headers
are now part of their own `_Builtin_XXX` modules.

See:
- https://github.com/apple/llvm-project/pull/7618
- https://reviews.llvm.org/D159483
- https://reviews.llvm.org/D159064

kateinoigakukun committed:
[wasm][test] Update test/IRGen/section_wasm.swift

To follow up changes in 5f037efee37f881f11efc25b6aff14166ae81c01

kateinoigakukun commented:
Added a few test fix commits that are found after rebranch 🩹

kateinoigakukun commented:
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
@swift-ci Please smoke test

kateinoigakukun commented:
preset=buildbot_linux_crosscompile_wasm
@swift-ci Please test with preset Linux Platform

kateinoigakukun commented:
@swift-ci Please smoke test

MaxDesiatov commented on stdlib/public/Platform/wasi-libc.modulemap:
```diff
@@ -19,20 +19,12 @@ module SwiftWASILibc [system] {
   header "ctype.h"
   header "errno.h"
   header "fenv.h"
-  header "float.h"
```
Are these no longer needed or are they transitively included by other headers?

kateinoigakukun replied:
They are now part of `_Builtin_XXX` modules provided by clang https://reviews.llvm.org/D159064

MaxDesiatov reviewed (approved):
🤩 Super excited to see this, thank you!
